server/audit_math/minerva.py

```diff
@@ -14,14 +100,193 @@ def get_sample_size(
+
+    >>> c3 = make_arlo_contest({"a": 600, "b": 400, "c": 100, "_undervote_": 100})
+    >>> get_sample_size(10, c3, None, [])
+    {'0.7': {'type': None, 'size': 134, 'prob': 0.7}, '0.8': {'type': None, 'size': 166, 'prob': 0.8}, '0.9': {'type': None, 'size': 215, 'prob': 0.9}}
+    >>> get_sample_size(20, c3, None, [])
+    {'0.7': {'type': None, 'size': 87, 'prob': 0.7}, '0.8': {'type': None, 'size': 110, 'prob': 0.8}, '0.9': {'type': None, 'size': 156, 'prob': 0.9}}
+    >>> get_sample_size(10, c3, make_sample_results(c3, [[55, 40, 3]]), {1: 100})
+    {'0.9': {'type': None, 'size': 150, 'prob': 0.9}}
+    """
+
+    if sample_results is not None:
+        # Construct round schedule as a function of only first round size.
+        # and other information set at the start of the audit.
+
+        # Minerva is not designed to be this formulaic in its round sizes,
+        # but while we work on rigorously demonstrating that round sizes can be chosen freely
+        # leveraging information on sample results in previous rounds, we offer this
+        # approach which simply defines all round sizes uniformly based on the first
+        # round size.
+
+        first_round_size = round_sizes[1]
+        prev_round_count = len(round_sizes)
+        next_round_size = minerva_round_size(first_round_size, prev_round_count)
+        logging.debug(f"{round_sizes=}, {next_round_size=}")
+        return {"0.9": {"type": None, "size": next_round_size, "prob": 0.9}}
+
+    alpha = Decimal(risk_limit) / 100
+    assert alpha < 1, "The risk-limit must be less than one!"
+
+    quants = [0.7, 0.8, 0.9]
+
+    samples: Dict = {}
+
+    p_w = Decimal("inf")
+    p_l = Decimal(-1)
+
+    margin = contest.margins
+    # Get smallest p_w - p_l
+    for winner in margin["winners"]:
+        if margin["winners"][winner]["p_w"] < p_w:
+            p_w = Decimal(margin["winners"][winner]["p_w"])
 
-    raise Exception("Not implemented!")
+    for loser in margin["losers"]:
+        if margin["losers"][loser]["p_l"] > p_l:
+            p_l = Decimal(margin["losers"][loser]["p_l"])
+
+    # If we're in a single-candidate race, set sample to 0
+    if not margin["losers"]:
+        for quant in quants:
+            samples[str(quant)] = {"type": None, "size": -1.0, "prob": quant}
+
+        return samples
+
+    num_ballots = contest.ballots
+
+    # Handles ties
+    if p_w == p_l:
+        for quant in quants:
+            samples[str(quant)] = {"type": None, "size": num_ballots, "prob": quant}
+
+        return samples
+
+    # Handle landslides.   TODO: this works for a 10% risk limit. But remove this case if handled upstream
+    if p_w == 1.0:
+        samples["0.9"] = {
+            "type": None,
+            "size": 4,
+            "prob": 1.0,
+        }
+
+        return samples
+
+    # If we haven't returned yet, actually do the math for the first round
+    for quant in quants:
+        size = minerva_sample_sizes(alpha, p_w, p_l, 0, 0, quant)
```

Comment:
@nealmcb It occurs to me that maybe this isn't the right way to do this. Looking at the library example [here](https://github.com/filipzz/athena/blob/master/code/athena.ipynb) it seems like we could pass a list of quantiles to the Athena library directly, which might dramatically speedup sample size computation, since it won't have to recompute the whole distribution three to four times. Thoughts?